read-fonts/src/tables/postscript/scale.rs

```diff
@@ -0,0 +1,541 @@
+//! Scaler for CFF outlines.
+
+use std::ops::Range;
+
+use super::{
+    dict::{self, Blues},
+    BlendState, Error, FdSelect, Index,
+};
+use crate::{
+    tables::{cff::Cff, cff2::Cff2, variations::ItemVariationStore},
+    types::{F2Dot14, Fixed, GlyphId, Pen},
+    FontData, FontRead, TableProvider,
+};
+
+/// Type for loading, scaling and hinting outlines in CFF/CFF2 tables.
+///
+/// # Subfonts
+///
+/// CFF tables can contain multiple logical "subfonts" which determine the
+/// state required for processing some subset of glyphs. This state is
+/// accessed using the [`FDArray and FDSelect`](https://adobe-type-tools.github.io/font-tech-notes/pdfs/5176.CFF.pdf#page=28)
+/// operators to select an appropriate subfont for any given glyph identifier.
+/// This process is exposed on this type with the
+/// [`subfont_index`](Self::subfont_index) method to retrieve the subfont
+/// index for the requested glyph followed by using the
+/// [`subfont`](Self::subfont) method to create an appropriately configured
+/// subfont for that glyph.
+///
+/// # Example
+///
+/// ```
+/// # use read_fonts::{tables::postscript::*, types::*, FontRef};
+/// # fn example(font: FontRef, coords: &[F2Dot14], pen: &mut impl Pen) -> Result<(), Error> {
+/// let scaler = Scaler::new(&font)?;
+/// let glyph_id = GlyphId::new(24);
+/// // Retrieve the subfont index for the requested glyph.
+/// let subfont_index = scaler.subfont_index(glyph_id);
+/// // Construct a subfont with the given configuration.
+/// let size = 16.0;
+/// let coords = &[];
+/// let with_hinting = false;
+/// let subfont = scaler.subfont(subfont_index, size, coords, with_hinting)?;
+/// // Scale the outline using our configured subfont and emit the
+/// // result to the given pen.
+/// scaler.outline(&subfont, glyph_id, coords, pen)?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct Scaler<'a> {
+    version: Version<'a>,
+    top_dict: ScalerTopDict<'a>,
+    units_per_em: u16,
+}
+
+impl<'a> Scaler<'a> {
+    /// Creates a new scaler for the given font.
+    ///
+    /// This will choose an underyling CFF2 or CFF table from the font, in that
+    /// order.
```

Comment:
Globally or can we pull some glyphs from cff, some from cff2? ...is that even a legal setup...?

The example suggests the user must be very aware of the fonts internals. What should a noob do, can we have a Just Works scaler that doesn't expose so many of the internals? 

Think I "just" want to get a damn outline and I don't much care what outline table it came from.

Reply:
My assumption is that anyone using read-fonts directly will definitely need to have a solid understanding of font internals. To my mind, that’s the whole point of the library. Skrifa is supposed to be the “easy mode” API.

Reply:
Per IM, mix and match isn't plausible due to the dense nature of the charstring index (and loca table).